src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -210,11 +211,21 @@ public class List<T> : IList<T>, System.Collections.IList, IReadOnlyList<T>
         // before adding the new element.
         //
         public void Add(T item) {
-            if (_size == _items.Length) EnsureCapacity(_size + 1);
-            _items[_size++] = item;
+            int size = _size;
+            if (size == _items.Length) GrowCapacityForAdd();
+            _items[size] = item;
+            _size = size + 1;
             _version++;
         }
 
+        // Separated out of List.Add to improve its code quality
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void GrowCapacityForAdd()
```

Comment:
>  where caller would get inlined regardless if it was extremely simple.

This sounds like a bug that we should fix if you still see it. NoInlining should always mean NoInlining, no exceptions.